Add periodically resetting meter option to utility meter

Wesley-Vos opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
The Utility Meter has now an additional option to set whether the source sensor is periodically resetting or not.

The source sensor of the utility meter can be of different kinds, ones who may periodically reset (for example a smart plug at boot) or ones that will never (for example the domestic utility meter that never resets, until you physically replace it).

Sometimes, the source sensor may become unavailable which is a problem if you want an accurate utility meter reading. Given that you don’t know much about the actual source sensor, the utility meter will only try to calculate the difference between a new state and old state (if delta values is not used) and add that to the current state of the utility meter. This mechanism is the safest option, but could be loosened a bit in the scenario that your sensor never resets, because you can then use the last valid state instead of the previous state.

Therefore, to make use of this scenario, an option is added to indicate that the meter is not periodically resetting and therefore always calculate the difference between the new valid state and last (not necessarily previous) state.

This option is by default enabled so that it will not break the current implementation. This new feature could be turned on (i.e. set the option off) by either adding `periodically_resetting: false` to a configuration.yaml or change an existing utility meter through the options flow.

The last valid state is always set upon receiving a new valid state and restored on start of Home Assistant.

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #72943
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26306

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Wesley-Vos committed:
Use last valid state if meter is not periodically resetting

Wesley-Vos committed:
Fix unload of entry, used during options flow submit

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -355,53 +377,70 @@ def start(self, unit):
+        new_state: State | None = event.data.get("new_state")
 
-        if self._state is None and new_state.state:
+        if new_state and self._state is None:
```
shouldn't it be `new_state.state` ?

and why change this line ?

Wesley-Vos replied:
The current `if ... and new_state.state` checks the state of the new_state State object whether it is None or not. However, the state attribute of a State object is always a string and therefore never None. If the actual reason for this line is to check if there is a valid new state value, we should maybe move lines 426-436 up, because that line is actually checking if the state value is valid.

I thought that the check was just there to verify if there is a new state in the Event and if so, start the meters (even when the state itself is invalid). Just let me know what is should be. I think that there is something to say for both situations, given that the result is the same.

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -355,53 +377,70 @@ def start(self, unit):
-            return
-
-        self._unit_of_measurement = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+                sensor.start(new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT))
```
why are you dropping source_state for new_state ?

Wesley-Vos replied:
I'm dropping the `source_state` for the `new_state` because the result of `hass.states.get(source_entity)` is the state we received through the Event. Furthermore, if we keep using the `source_state`, we need to implement an additional None check (because the state can be None).

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -355,53 +377,70 @@ def start(self, unit):
+    def calculate_adjustment(
+        self, old_state: State | None, new_state_val: Decimal
+    ) -> Decimal | None:
+        """Calculate the adjustment based on the old and (assumed valid) new state."""
```
based on the description of the method, maybe they should both be `State` as in `new_state.value` ?(just trying to make things clearer)

Wesley-Vos replied:
Yes and no, in order to get both values, you have to parse both states as a decimal while you might not need both values (in the case of delta values or the new periodically resetting option), which might raise DecimalExceptions if the old state is invalid.

I agree that both argumetns should be of the same type to make it clearer, I will think about another solution.

Wesley-Vos replied:
Yes true, but if the delta_values are used, we are only interested in the new_state and not in the old state. If we, before calling this calculate_adjustment method already check if both the old and new state are valid (because that is needed in order to get the Decimal value of a state), we might add an error log that the old state is invalid while we are not even interested in the old state.

Another option is to do Decimal parsing for both the old_state and new_state in the `calculate_adjustment` method but than we cannot already return once we discover that the new_state is invalid and continuing is useless, as I do now.

dgomes replied:
check if it's delta or not and parse to Decimal accordingly in `calculate_adjustment` ?

Wesley-Vos replied:
I've changed both parameters to the `State` object. The old_state will, if necessary be parsed inside `calculate_adjustment` wrapped in a try/ except to catch parsing errors. The new_state is parsed inside `calculate_adjustment` as well including a try/ except. While this new_state will never fail to parse (because it is already validated before the `calculate_adjustment` is called), it is safer for the future, if the `calculate_adjustment` may be re-used somewhere else.

dgomes reviewed (commented):
Left some comments

gapple commented:
I'm not too familiar yet with either HA or Python, so I may be incorrect in what I understand from the code: 
- it seems like this will help with more accuracy after restart or when the sensor is otherwise not available for a period (when there is no `old_state` to compare to, but now there will be a more persistent `last_valid_state`)
- but if the sensor resets to 0 or rolls over at a maximum value, then it will not calculate the difference from zero to the first reading - it will only reset the `last_valid_state` to a new value; the second reading after reset will then calculate the difference from the first reading.

If `new_state < last_valid_state`, then it seems like returning `new_state` as the adjustment would be more accurate.

For #31404, the reset point is a known value (e.g. an unsigned 32-bit integer), so `rollover_value - last_valid_state + new_state` would better capture the change since the last reading if that rollover value can be configured as well.

emontnemery commented:
I'm marking the PR as draft based on the ongoing discussion. Please mark the PR as ready for review when the discussion has been resolved 👍

Wesley-Vos committed:
Adjustments based on code review

Wesley-Vos commented:
> I'm not too familiar yet with either HA or Python, so I may be incorrect in what I understand from the code: 
> 
> - it seems like this will help with more accuracy after restart or when the sensor is otherwise not available for a period (when there is no `old_state` to compare to, but now there will be a more persistent `last_valid_state`)
> 
> - but if the sensor resets to 0 or rolls over at a maximum value, then it will not calculate the difference from zero to the first reading - it will only reset the `last_valid_state` to a new value; the second reading after reset will then calculate the difference from the first reading.
> 
> 
> 
> If `new_state < last_valid_state`, then it seems like returning `new_state` as the adjustment would be more accurate.
> 
> 
> 
> For #31404, the sensor value resets at known value but is otherwise persistent (e.g. an unsigned 32-bit integer), so `rollover_value - last_valid_state + new_state` would better capture the change since the last reading if that rollover value can be configured as well.

Hi! Yes you are correct about your first observation. This fix will create more accuracy after the source sensor is unavailable as well as after a restart. The first is the main fix, the second one (after restart) is just a side-effect. It looks like that the first state change of the source sensor is not registered by the utility meter after HA starts which leads to a meter reading miss. However, if the last_valid_state is used, this issue is fixed as well because the missed adjustment is part of the next state change and adjustment. Maybe, the issue about missed readings after restart has to be further investigated separately because it looks like it is there.

It is true that if the sensor rolls over or reset, the adjustment from 0 to any value might be missed in some cases (for example if it goes from overflow to a non-zero state. If it goes from overflow to 0, everything is still fine because 0 to non-zero is captured). Given that we don't know anything about the source sensor, I'm not sure if we should assume that it resets to 0, what if it resets to 1 and the first new state is 1.5. Then there is no way to know whether we should add 1.5 or only 0.5.

Adding a configurable rollover point is quite advanced and a I personally think a rare use-case. Furthermore, if this is implemented, the utility meter is fixing 'strange' behaviour of the source sensor which should not be the case. The strange behaviour should be fixed further down, somewhere close to the sensor, hope that makes sense?

Thanks for the feedback!

Wesley-Vos commented:
@dgomes I've adjusted the code based on your comments, thanks! I think that it is ready for review :)

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -387,9 +387,17 @@ def _validate_state(state: State | None) -> Decimal | None:
+        try:
+            if (new_state_val := self._validate_state(new_state)) is None:
+                return None
+        except DecimalException as err:
```
Why not validate that it can be cast to Decimal in _validate_state (Returning None if not valid)

Wesley-Vos replied:
I chose to not do the Exception handling inside `_validate_state` because the error message you want to log depends on the state you are validating, If the new state is invalid, you may want to indicate the state is invalid, while if the new_state is valid but the old one is invalid (and needed), than you want to log both states and show that it cannot be used.

If the logger is already inside `_validate_state`, you will first receive an error that the state is invalid, if the new state is invalid, and receive the same error in the next run, because then, the old state is still invalid and will trigger a new error log.

Another option I'm thinking about is to do the Exception handling inside the validate_state method and then if it fails to parse the Decimal, we return None and throw an appropriate error log in the scope we called the method, i.e. something like:
```python
if (new_state_val := self._validate_state(new_state)) is None:
    _LOGGER.error("Invalid state %s", new_state.state)
   return
```

Do you agree with this solution?

dgomes replied:
That is what I was proposing :)

Wesley-Vos replied:
Great, will change it! :)

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -413,34 +421,32 @@ def calculate_adjustment(
+            if (new_state_val := self._validate_state(new_state)) is None:
                 return
         except DecimalException as err:
             if self._sensor_delta_values:
```
We can probably get ride of this if and have a single warning message ?

Wesley-Vos replied:
Yes agree, I will combine them

Wesley-Vos replied:
Yes I agree, will combine them

dgomes commented:
Please attend coverage

Wesley-Vos committed:
Move DecimalException handling to validation method

Wesley-Vos commented:
> Please attend coverage

The coverage should be fine now that the exception handling has been moved back.

Wesley-Vos committed:
Add test for invalid new state in calculate_adjustment method

Wesley-Vos commented:
One line was not covered. As long as the `calculate_adjustment` method is not called outside the current scope it is not a problem. However, I've added a test to validate the specific situation, in case the method is called from an other scope in the future. Local coverage check verifies that the coverage miss has been fixed.

dgomes commented:
@Wesley-Vos The default configuration should have been False to maintain previous behaviour, now shouldn't it ?

https://github.com/home-assistant/core/issues/90930
